### PR TITLE
API-server audit log filtering policy in ClusterLogForwarder

### PR DESCRIPTION
**Pull request by alanconway**
/cc @jcantril
/cc @nmalik
/cc @eparis

### CODE REVIEW

File: enhancements/cluster-logging/api-audit-log-policy.md
Comment (jewzaam) [inline enhancements/cluster-logging/api-audit-log-policy.md:16]: Is there an enhancement to add this?  I can't find it if yes.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:34]:
```suggestion
- Allow audit logs consumers with different requirements to get the subset logs they need.
```

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:35]:
```suggestion
- Forward API server audit logs from the management cluster, as they are not available in the hosted cluster.
```
This is saying forward to the configured destination, not that we must forward to the data plane as an intermediate?

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:37]: I don't see centralized forwarding necessarily as a motivation.  It's an optimization and an implementation detail.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:41]:
```suggestion
#### As a service consumer, I want to forward selected API audit logs to my chosen destinations.
```

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:43]:
```suggestion
- configure API audit policy and forwarding using logging customer resources in the hosted cluster.
```

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:47]:
```suggestion
#### As a service provider, I want to forward selected API audit logs to my chosen destinations.
```

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:48]:
```suggestion
- configure the SD internal audit policy
- configure API audit policy and forwarding using namespaced logging customer resources in the management cluster.
```

Calling out namespaced resources here as we must keep in mind that management clusters _are_ OSD clusters right now.  The audit forwarding of the management cluster should leverage the non-hypreshift solution documented in this PR.  The MC is hosting many HCP.  These must be forwarded as well and the policy and configuration of outputs must be independent of the MC configuration.  For example, we have different types of events possibly and we ship to a different index in splunk.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:49]:
```suggestion
  - for example: Red Hat Service Delivery team forwards audit logs to internal splunk destination.
```

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:50]: This is to say configuration is independent of the customer (service consumer) configuration in the hosted cluster?

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:54]:
```suggestion
- Multiple audit policies can be defined in a management cluster, independent of hosted cluster policies.
- Multiple audit policies can be defined in a hosted cluster, independent of the management cluster policies.
```

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:56]:
```suggestion
- Control log forwarding from the _management_ cluster with configuration in the _hosted_ cluster.
```

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:58]:
```suggestion
  - hosted cluster configuration requests forwarding, management cluster controls what is delivered.
```

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:63]:
```suggestion
- May not support every log destination or configuration supported by CLO today.
```

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:64]: Is this restriction enforced by what is exposed in `HypershiftLogForwarder` or is there some other configuration option available to the service provider to further restrict functionality?

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:108]: The functionality needed on the MC is captured in the `HypershiftLogForwarder`.  Could that CR be used in each HCP NS to aggregate everything?

1. SD wites a HLF to the HCP NS with the audit forwarding needed to splunk.  
2. Customer writes HLF in the hosted cluster with the customer forwarding needs.
3. HLO watches the hosted cluster HLF and adds to the HLF already created, in the status.  If HLF doesn't exist, one is created with no spec.

Something like this would mean CLF isn't changed, no breaking API changes to deal with.  All the configuration ends up in one place.  It's local to the HCP, so authentication (STS) is fine.  No need for transformation, the CRDs will be the same in the management cluster and hosted cluster.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:156]: Sure but you need a webhook still per HCP to recieve the events and write to a filesystem.  If you do a sidecar we have problems of order of containers starting in the pod.  This is solved with the webhook provided by kube.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:158]: Yes, but this is privileged as it requires access to kube credentials for each HCP and cross NS permissions to access each HCP.  Before doing this it needs run by SRE security, prodsec, compliance.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:160]: In what ways would we need to restrict destinations?  The limitations will need to be on what network paths are available from the management cluster.  The MC can talk to the destination directly.  Configuration of any network paths by the service consumer to facilitate this must be out of scope of this enhancement.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:165]: Recieving events from the HCP APIs will be via webhook specific to that HCP.  After that is processed it's up to how we architect this solution with respect to how CLO receives the events.

The challenge is when we look at authentication to various destinations.  The solution needs to support things like CloudWatch for ROSA HCP.  This requires STS, which requires a token per HCP.  Tokens expire after 60 minutes.  CLO will need to refresh tokens from each HCP.  We have code for this in the current solution (on gitlab) that can be referenced.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:189]: It's not new depending on who you ask :) 
SRE have several operators running in each HCP and there are the HCP components themselves, which run as pods on the MC and connect to the HCP API.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:195]:
```suggestion
The custom Splunk forwarder operator and splunk audit exporter webhook will be removed.
```

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:213]: Phase 2 of the work will enable self service.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:216]: But adds the ability to filter, which is critical for our use case.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:217]:
```suggestion
- Was rushed to add support for Cloudwatch
```

**Review by jewzaam — CHANGES_REQUESTED** [review]
Things we need to dig into, and are called out but it's worth emphasis, are how HLO and HLF are deployed / managed and how authentication is managed for STS.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (alanconway) [reply enhancements/cluster-logging/hypershift-audit-logs.md:161]: This is so the HLO on the management cluster to reconcile the HypershiftLogForwarders on the hosted clusters.
Agreed it needs to be secured properly, that's an issue with CLO in general as it has access to a lot of potentially sensitive logs.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (alanconway) [reply enhancements/cluster-logging/hypershift-audit-logs.md:163]: It is very possible there are no restrictions needed. I had a vague idea that there might be connection issues from "different" clusters, but you are correct that if the destination is reachable from the MC then there's no reason we can't forward to it.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (alanconway) [reply enhancements/cluster-logging/hypershift-audit-logs.md:168]: +1. The CLF resource references all the relevant credentials and connection info, the HLF resource would do the same. See other proposal: The HLO on the MC would use the HLF from the hosed cluster as a template to construct a CLF in the HCP that does the actual forwarding.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (alanconway) [reply enhancements/cluster-logging/hypershift-audit-logs.md:192]: Glad to hear it! I felt like it should be possible but didn't know if it had been done.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (alanconway) [reply enhancements/cluster-logging/hypershift-audit-logs.md:219]: Agreed - and to customer use cases also. Hence the goal of transplanting it to the CLO where all can benefit.

File: enhancements/cluster-logging/api-audit-log-policy.md
Comment (jewzaam) [reply enhancements/cluster-logging/api-audit-log-policy.md:97]: Makes sense, I like apiAuditPolicyRef

File: enhancements/cluster-logging/api-audit-log-policy.md
Comment (jewzaam) [reply enhancements/cluster-logging/api-audit-log-policy.md:132]: Easiest thing I could do is grab aggregate data from a management cluster.  If we need more let me know but the numbers are pretty small.

Total # of HCP = 51

CPU
- min = 0m
- avg = 0.6m
- max = 4m

Memory
- min = 22Mi
- avg = 47.5Mi
- max = 64Mi

**Comment by jewzaam** [comment]
@alanconway I didn't see any mention of editing the payload that is shipped to an output, is that in scope?  For example, if SRE has audit policy set to something other than "Default", can a customer remove the request/response data before shipping?

**Comment by alanconway** [comment]
> @alanconway I didn't see any mention of editing the payload that is shipped to an output, is that in scope? For example, if SRE has audit policy set to something other than "Default", can a customer remove the request/response data before shipping?

Yes, that's what the kube config file allows you to specify. The idea here is to use one or more _filters_ at the cluster logging operator, not to re-configure the API server itself. That is what will allow separate audit streams to each party. The API server would have to be logging everything at full volume, so each party can filter as desired. I believe that's currently the case for classic ROSA clusters (full volume at API server, separate filter process based on kube policy)  and hasn't been a performance issue to date.

**Comment by alanconway** [comment]
@jewzaam can you give me approved and/or lgtm? Any outstanding concerns?

**Comment by alanconway** [comment]
@eparis approval?

**Review by jewzaam — CHANGES_REQUESTED** [review]

File: enhancements/cluster-logging/api-audit-log-policy.md
Comment (bergmannf) [inline enhancements/cluster-logging/api-audit-log-policy.md:108]: The audit-exporter supports more than just applying the policy - so exposing these features should either explicitly be not part of this enhancement or there must be a way these can be specified: e.g.

```
     --allowed-keys strings                 allowed annotations (default [authorization.k8s.io/decision,authorization.k8s.io/reason,managed.openshift.io/rate-limited])
      --annotations stringToString           annotations to be added to each event (--annotations=my-annotation=my-value) (default [])
      [...]
      --dedupe                               deduplicate repeated update requests (default true)
      --dedupe-cache-ttl duration            max ttl for dedupe write cache (default 48h0m0s)
      --filter-status-codes int32Slice       ignore events with these http response codes (default [404,409,422,429])
      [...]
      --rate-limit                           enable rate limiting (default true)
      --rate-limit-burst int                 per-bucket burst (default 5)
      --rate-limit-count int                 rate limit count per period (default 1)
      --rate-limit-period duration           rate limit period (default 15m0s)
```

**Comment by alanconway** [comment]
@jewzaam I've incorporated all the latest comments, I think this is much clearer now. Take a look and /lgtm if there's nothing new. Who can /approve this? @eparis ?

File: enhancements/cluster-logging/api-audit-log-policy.md
Comment (jewzaam) [inline enhancements/cluster-logging/api-audit-log-policy.md:197]: The CLO installation is shared.  SRE doesn't install CLO right now so there are 2 scenarios:
1. Customer has CLO installed already.
2. Customer does not have CLO already installed.

For the first case we have an issue.  Quite simply, CLO is a customer workload and we will not and cannot manage it.  This is a major operational hurdle and was discussed at length in the context of CMA/KEDA for auto-scaling ingress.  It was decided that we could not support an SRE deployment of CMA.  This is a material risk for the adoption of this solution within SD and should be called out that acceptance of this enhacnement does not guarentee consumption by SD.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:94]: What does this mean?  Is it that CLO is watching all namespaces?

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:96]: I read "single" as "singleton instance".  This means we have a single template for forwarding configuration that must be shared across all HCP on that management cluster.  We need to make this namespace scoped.  Deployment of that resource is a concern for the service provider.  It provides flexibility at a HCP level.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:115]: The name is not intuitive even if it's a Global scope resource.  It's a template, so let it be called a template.

```suggestion
#### ClusterLogForwarderTemplate
```

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:113]: 1. what are the base set of output types that is supported in HLF?
2. how does the service provider restrict the supported output types?

SRE needs splunk as an output.  If we know this is in scope I would like it listed so that is clear.  Also from customers I am hearing CloudWatch and syslog.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:195]: How can this be true if CLF is moving from global to namespace scope?

**Review by jewzaam — CHANGES_REQUESTED** [review]

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (alanconway) [reply enhancements/cluster-logging/hypershift-audit-logs.md:195]: The CLF is already in namespace scope, but restricted to only deploy in the "openshift-logging" namespace with the name "instance", so there can be only one instance. We just need to lift that restriction and fix assumptions in the code about the constant namespace and name.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (alanconway) [reply enhancements/cluster-logging/hypershift-audit-logs.md:167]: Will add a note on STS. 
I don't think we need to restrict output types available for the HLF initially. I noted early on that we _could_ enforce  restrictions if there was some case that was problematic, but so far I'm not aware of any. We can start by allowing everything and add restrictions if/when we need to.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (alanconway) [reply enhancements/cluster-logging/hypershift-audit-logs.md:113]: I will remove the talk of out restrictions, if it turns out we do need such restrictions we can introduce them later. 
It was an idea I had at the beginning but it doesn't seem to be really necessary.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (alanconway) [reply enhancements/cluster-logging/hypershift-audit-logs.md]: Fixed - instance per HCP as discussed.

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (alanconway) [reply enhancements/cluster-logging/hypershift-audit-logs.md]: Yep - the CLF is namespace-scoped

File: enhancements/cluster-logging/api-audit-log-policy.md
Comment (alanconway) [reply enhancements/cluster-logging/api-audit-log-policy.md:205]: The customers CLO is deployed on the hosted cluster. SRE's instance is deployed on the management cluster. They are fully separated. The customer has no _direct_ way of influencing the SRE's CLO.
The customer can create a HyperShiftLogForwarder. The HLO on the management cluster combines this with the service providers ClusterLogForwarderTemplate to produce a ClusterLogForwarder on the HCP. This contains config pulled from the hosted HLF, but otherwise it is entirely controlled by the CLO on the management cluster.
So SD has full control over everything running on the management cluster. The management CLO will run collectors on the management cluster, but cost-wise that's no different from running the existing splunk exporter.

File: enhancements/cluster-logging/api-audit-log-policy.md
Comment (alanconway) [reply enhancements/cluster-logging/api-audit-log-policy.md:201]: True, but this is limited to configuration in the customers HLF, which are limited to forwarding audit logs.
There's a clear distinction: 
- The customer can install any version of CLO they like, and use any features they like in their CLF configurations (all logs other than API-audit logs). 
- For API audit logs, the customer must create a HLF based on the version of CLO used by SD.
So SD doesn't have to match the customer version, but does need to advertize the version they use, since that's the version for HLF resources written by the customer.  Potentially confusing but not fatal.

Comment (alanconway) [reply enhancements/cluster-logging/api-audit-log-policy.md:201]: The text is badly worded: "The same CLO operator will manage the customer and SRE log forwarding.", what I meant to say is "the CLO on the managed server will control log forwarding for SD and for customer API-audit logs (only)" The managed cluster is taking on some user work but only for API-audit logs, and that because we have no choice.

Comment (alanconway) [reply enhancements/cluster-logging/api-audit-log-policy.md:201]: fixed the text

File: enhancements/cluster-logging/api-audit-log-policy.md
Comment (jewzaam) [inline enhancements/cluster-logging/api-audit-log-policy.md:213]: Contents of this section should be merged into hypershift-audit-logs.md.  Not sure what would be appropriate here in api-audit-log-policy.md.

File: enhancements/cluster-logging/api-audit-log-policy.md
Comment (jewzaam) [reply enhancements/cluster-logging/api-audit-log-policy.md:217]: Context of this comment is not hypershift-audit-logs.md, in classic OSD/ROSA there is no management cluster involved.  There is one CLO on the customer's cluster.

**Comment by alanconway** [comment]
@jewzaam @eparis @cahartma @bergmannf  can I get an /lgtm /approved on this please?

File: enhancements/cluster-logging/api-audit-log-policy.md
Comment (jewzaam) [inline enhancements/cluster-logging/api-audit-log-policy.md:132]: Move this section to hypershift-audit-logs.md.  Any considerations for STS for non-hypershift ROSA?

File: enhancements/cluster-logging/hypershift-audit-logs.md
Comment (jewzaam) [inline enhancements/cluster-logging/hypershift-audit-logs.md:70]:
```suggestion
  - management cluster HLO automatically validates data-plane requests, and enables them if they are supported.
```

**Review by jewzaam — CHANGES_REQUESTED** [review]

File: enhancements/cluster-logging/api-audit-log-policy.md
Comment (alanconway) [reply enhancements/cluster-logging/api-audit-log-policy.md:132]: None that I know of, we already have STS support there. Again got my wires crossed.

**Comment by alanconway** [comment]
@jewzaam took  your latest suggestions, no new changes.  LGTM?

**Comment by alanconway** [comment]
@jewzaam hopefully final draft..

**Pull request merged**
